File: algorithms/js/120.三角形最小路径和.js
/*
 * @lc app=leetcode.cn id=120 lang=javascript
 *
 * [120] 三角形最小路径和
 */

// @lc code=start
/**
 * @param {number[][]} triangle
 * @return {number}
 */
var minimumTotal = function(triangle) {
  const row = triangle.length;
  const col = triangle[row - 1].length;
  const dp = Array.from({ length: row + 1 }, (_, i) =>
    Array.from({ length: col + 1 }, () => (i === 0 ? 0 : Number.MAX_VALUE))
  );
  for (let i = 1; i < row + 1; i++) {
    for (let j = 1; j < triangle[i - 1].length + 1; j++) {
      dp[i][j] =
        triangle[i - 1][j - 1] + Math.min(dp[i - 1][j], dp[i - 1][j - 1]);
    }
  }
  return Math.min(...dp[row]);
};
// @lc code=end

File: algorithms/golang/120.三角形最小路径和.go
/*
 * @lc app=leetcode.cn id=120 lang=golang
 *
 * [120] 三角形最小路径和
 */

// @lc code=start

/**
 * dp, bottom-top
 * 时间复杂度：O(n^2)，空间复杂度：O(n)
 */
func minimumTotal(triangle [][]int) int {
	size := len(triangle)
	dp := make([]int, size+1)
	min := func(a, b int) int {
		if a < b {
			return a
		} else {
			return b
		}
	}
	for i := size - 1; i > -1; i-- {
		for j := 0; j < len(triangle[i]); j++ {
			dp[j] = min(dp[j], dp[j+1]) + triangle[i][j]
		}
	}
	return dp[0]
}

// @lc code=end

